Language/Haskell/GhcMod/ImportedFrom.hs

```diff
@@ -0,0 +1,785 @@
+
+getVisibleExports
+    :: forall m. (GhcMonad m, MonadIO m, GmOut m, GmLog m)
+    => (String -> m (Maybe String))
+    -> String
+    -> m (Maybe (M.Map String [String]))
+getVisibleExports getHaddockInterfaces p = do
+    gmLog GmDebug "" $ strDoc $ "getVisibleExports: " ++ p
+
+    let p' = case splitOn "@" p of
+                [p0, _] -> p0
+                _       -> p
+    
+    haddockInterfaceFile <- getHaddockInterfaces p'
+    join <$> traverse getVisibleExports' haddockInterfaceFile
+
+  where
+
+    getVisibleExports' :: forall m. (GhcMonad m, MonadIO m, GmOut m, GmLog m) => FilePath -> m (Maybe (M.Map String [String]))
+    getVisibleExports' ifile = do
+        iface <- Haddock.readInterfaceFile nameCacheFromGhc ifile
+
+        case iface of
+            Left _          -> do gmErrStrLn $ "Failed to read the Haddock interface file: " ++ ifile
+                                            ++ "You probably installed packages without using the '--enable-documentation' flag."
+                                            ++ ""
+                                            ++ "Try something like:\n\n\tcabal install --enable-documentation p"
+                                  error "No haddock interfaces file, giving up."
+            Right iface'    -> do let m  = map (\ii -> (Haddock.instMod ii, Haddock.instVisibleExports ii)) $ Haddock.ifInstalledIfaces iface' :: [(Module, [Name])]
+                                      m' = map (\(mname, names) -> (showSDoc tdflags $ ppr mname, map (showSDoc tdflags . ppr) names)) m       :: [(String, [String])]
+                                  return $ Just $ M.fromList m'
+
+    ------------------------------------------------------------------------------------------------------------------------
+    -- Copied from http://hackage.haskell.org/package/haddock-api-2.16.1/docs/src/Haddock-InterfaceFile.html#nameCacheFromGhc
+    -- but for a general monad m instead of the specific monad Ghc.
+    ------------------------------------------------------------------------------------------------------------------------
+    nameCacheFromGhc :: forall m. (GhcMonad m, MonadIO m, GmOut m, GmLog m) => Haddock.NameCacheAccessor m
+    nameCacheFromGhc = ( read_from_session , write_to_session )
+      where
+        read_from_session = do
+           ref <- GhcMonad.withSession (return . hsc_NC)
+           liftIO $ readIORef ref
+        write_to_session nc' = do
+           ref <- GhcMonad.withSession (return . hsc_NC)
+           liftIO $ writeIORef ref nc'
+
+-- | Convert a module name string, e.g. @Data.List@ to @Data-List.html@.
+moduleNameToHtmlFile :: String -> String
+moduleNameToHtmlFile m =  map f m ++ ".html"
+    where f :: Char -> Char
+          f '.' = '-'
+          f c   = c
+
+filterMatchingQualifiedImport :: String -> [NiceImportDecl] -> [NiceImportDecl]
+filterMatchingQualifiedImport symbol hmodules =
```

Comment:
Also should be in a `where`, the indenting looks strange also. Newline after `of` please.